hippylib/modeling/prior.py

```diff
@@ -604,4 +605,123 @@ def sample(self, noise, s, add_mean=True):
         self.Asolver.solve(s, rhs)
         
         if add_mean:
-            s.axpy(1., self.mean)
\ No newline at end of file
+            s.axpy(1., self.mean)
+
+
+class GaussianRealPrior(_Prior):
+    """
+    This class implements a finite-dimensional Gaussian prior,
+    :math:`\\mathcal{N}(\\boldsymbol{m}, \\boldsymbol{C})`, where
+    :math:`\\boldsymbol{m}` is the mean of the Gaussian distribution, and
+    :math:`\\boldsymbol{C}` is its covariance. The underlying finite element
+    space is assumed to be the "R" space.
+    """
+
+    def __init__(self, Vh, covariance, mean=None):
+        """
+        Constructor
+
+        Inputs:
+        - :code:`Vh`:             Finite element space on which the prior is
+                                  defined. Must be the Real space with one global 
+                                  degree of freedom
+        - :code:`covariance`:     The covariance of the prior. Must be a
+                                  :code:`numpy.ndarray` of appropriate size
+        - :code:`mean`(optional): Mean of the prior distribution. Must be of
+                                  type `dolfin.Vector()`
+        """
+
+        self.Vh = Vh
+
+        if Vh.dim() != covariance.shape[0] or Vh.dim() != covariance.shape[1]:
+            raise ValueError("Covariance incompatible with Finite Element space")
+
+        self.covariance = covariance
+        
+        #np.linalg.cholesky automatically provides error checking, so use those
+        self.chol = np.linalg.cholesky(self.covariance)
+
+        self.chol_inv = scila.solve_triangular(
+                                        self.chol,
+                                        np.identity(Vh.dim()),
+                                        lower=True)
+
+        self.precision = np.dot(self.chol_inv.T, self.chol_inv)
+
+        trial = dl.TrialFunction(Vh)
+        test  = dl.TestFunction(Vh)
+        
+        domain_measure_inv = 1.0 / dl.assemble(dl.Constant(1.) * dl.dx(Vh.mesh()))
+        #mass matrix
+        self.M = dl.assemble(domain_measure_inv * dl.inner(trial, test) * dl.dx)
+        self.Msolver = Operator2Solver(self.M)
+
+        if mean:
+            self.mean = mean
+        else:
+            tmp = dl.Vector()
+            self.M.init_vector(tmp, 0)
+            tmp.zero()
+            self.mean = tmp
+
+        if Vh.dim() == 1:
+
+            trial = dl.as_matrix([[trial]])
+            test  = dl.as_matrix([[test]])
+
+        #Create form matrices 
+        covariance_form = dl.as_matrix(list(map(list, self.covariance)))
+        precision_form  = dl.as_matrix(list(map(list, self.precision)))
+        chol_form       = dl.as_matrix(list(map(list, self.chol)))
+        chol_inv_form   = dl.as_matrix(list(map(list, self.chol_inv)))
+
+        #variational for the regularization operator, or the precision matrix
+        var_form_R = domain_measure_inv \
+                     * dl.inner(test, dl.dot(precision_form, trial))
```

Comment:
Unfortunately no. 

In the 1D case, `precision_form` is a tensor form of shape `(1,1)`. `precision_form * trial` results in another tensor form of shape `(1,1)`. Then, trying `dl.inner(test, precision_form * trial)` fails, because UFL thinks we're trying to compute the inner product of a scalar and a tensor form. `dl.dot(test, precision_form * trial)` also fails because the dot product of a scalar and tensor form isn't defined.

The problem arises because UFL can't recognize that a `(1,1)` tensor form is a scalar, so I'm forced to elevate `trial` and `test` to vector forms.